tpm2/structures.go

```diff
@@ -74,15 +51,264 @@ type tpmsPCRSelection struct {
 }
 
 // Public contains the public area of an object.
-//
-// TODO(awly): this is RSA-specific right now. Make it work for other Types.
 type Public struct {
 	Type       Algorithm
 	NameAlg    Algorithm
-	Attributes uint32
+	Attributes KeyProp
 	AuthPolicy []byte
-	Parameters tpmsRSAParams
-	Unique     []byte
+
+	// Only one of the Parameters fields should be set. When encoding/decoding,
+	// one will be picked based on Type.
+	RSAParameters *RSAParams
+	ECCParameters *ECCParams
+}
+
+func (p Public) encode() ([]byte, error) {
+	head, err := tpmutil.Pack(p.Type, p.NameAlg, p.Attributes, p.AuthPolicy)
+	if err != nil {
+		return nil, err
+	}
+	var params []byte
+	switch p.Type {
+	case AlgRSA:
+		params, err = p.RSAParameters.encode()
+	case AlgECC:
+		params, err = p.ECCParameters.encode()
+	default:
+		err = fmt.Errorf("unsupported type in TPMT_PUBLIC: %v", p.Type)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return concat(head, params)
+}
+
+func decodePublic(in *bytes.Buffer) (Public, error) {
+	var pub Public
+	var err error
+	if err = tpmutil.UnpackBuf(in, &pub.Type, &pub.NameAlg, &pub.Attributes, &pub.AuthPolicy); err != nil {
+		return pub, fmt.Errorf("decoding TPMT_PUBLIC: %v", err)
+	}
+
+	switch pub.Type {
+	case AlgRSA:
+		pub.RSAParameters, err = decodeRSAParams(in)
+	case AlgECC:
+		pub.ECCParameters, err = decodeECCParams(in)
+	default:
+		err = fmt.Errorf("unsupported type in TPMT_PUBLIC: %v", pub.Type)
+	}
+	return pub, err
+}
+
+// RSAParams represents parameters of an RSA key pair.
+//
+// Symmetric and Sign may be nil, depending on key Attributes in Public.
+// Modulus must always be non-nil.
+type RSAParams struct {
+	Symmetric *SymScheme
+	Sign      *SigScheme
+	KeyBits   uint16
+	Exponent  uint32
+	Modulus   *big.Int
+}
+
+func (p *RSAParams) encode() ([]byte, error) {
+	if p == nil {
+		return nil, nil
+	}
+	sym, err := p.Symmetric.encode()
+	if err != nil {
+		return nil, err
+	}
+	sig, err := p.Sign.encode()
+	if err != nil {
+		return nil, err
+	}
+	if p.Modulus == nil {
+		return nil, errors.New("RSAParams.Modulus must be set")
+	}
+	rest, err := tpmutil.Pack(p.KeyBits, p.Exponent, p.Modulus.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	return concat(sym, sig, rest)
+}
+
+func decodeRSAParams(in *bytes.Buffer) (*RSAParams, error) {
+	var params RSAParams
+	var err error
+
+	params.Symmetric, err = decodeSymScheme(in)
```

Comment:
style nit: This line can merge with the next if statement, and the same is true for line 146

Reply:
Done